Added Fluent API

RoyGoode opened the pull request:
The changes made in this pull request allow tables to be created using a Fluent API with attribute-free objects (aka POCOs). This allows for cleaner separation of layers and aids dependency injection by not requiring references to sqlite-net to decorate model classes.

Example usage:

```
connection.BuildMapping<SampleEntity>()
		  .TableName("ExampleTable")
		  .PrimaryKey(x => x.Key)
		  .CreateTable();

connection.BuildMapping<SampleEntity2>()
		  .TableName("Users")
		  .PrimaryKey(x => x.Id)
		  .AutoIncrement(x => x.Id)
		  .ColumnName(x => x.PasswordHash, "Password")
		  .Unique(x => x.Username)
		  .CreateTable();

connection.BuildMapping<SampleEntity3>()
		  .TableName("Contacts")
		  .Index("ix_full", x => x.Name, x => x.Email, x => x.Telephone)
		  .PrimaryKey(x => x.Id, true)
		  .MaxLength(x => x.Email, 255)
		  .NotNull(x => x.Email)
		  .Ignore(x => x.SecretMessage)
		  .CreateTable();
```
Based on the following example plain model classes...

```
public class SampleEntity
{
	public string Key { get; set; }

	public string Value { get; set; }
}

public class SampleEntity2
{
	public int Id { get; set; }

	public string Username { get; set; }

	public string PasswordHash { get; set; }
}

public class SampleEntity3
{
	public int Id { get; set; }

	public string Name { get; set; }

	public string Email { get; set; }

	public string Telephone { get; set; }

	public string SecretMessage { get; set; }
}
```

Note: CreateTable() must be called to trigger the creation of the mapping and resulting table. 

The additional code recycles existing functionality for creating table mappings from attributes.

Reza1024 commented:
I was looking forward to this! Removing decoration from the classes and moving the sqlite dependency to where it is being used will be awesome 👍

nberardi commented:
I made a change in #599 that is similar.  That caches and allows custom building of TableMappings.  I wonder if there is a way to combine the ideas.  See the `DefaultTableMapper`.  https://github.com/praeclarum/sqlite-net/pull/599/files?w=1#diff-e249207f2f9ac72b8e31b87d9be311e1R2523

The reason cached it was because I saw a lot of memory pressure, since most queries require the `TableMapping`.  The reason I created the `DefaultTableMapper` was because in a couple cases I needed to do custom mappings against tables, where my object structure needed to have a hierarchy but my tables needed to be flat.

praeclarum commented:
Hello!

I really love this but would like to change a few things too.

I think the builder should make `TableMappings` and those can be sent to `CreateTable`. This would make calling `CreateTablesAsync` more palatable.

So the syntax would be something like:

```csharp
var mapping = TableMapping.Build<SampleEntity3>()
		  .TableName("Contacts")
		  .Index("ix_full", x => x.Name, x => x.Email, x => x.Telephone)
		  .PrimaryKey(x => x.Id, true)
		  .MaxLength(x => x.Email, 255)
		  .NotNull(x => x.Email)
		  .Ignore(x => x.SecretMessage)
		  .ToMapping();
connection.CreateTable (mapping);
```

praeclarum commented on src/SQLite.cs:
```diff
@@ -338,6 +338,29 @@ public TableMapping GetMapping<T> ()
+		/// Adds a mapping to be used for indexes and naming for the given type.
+		/// </summary>
+		/// <param name="tableMapping">The table mapping.</param>
+		internal void AddMapping(TableMapping tableMapping)
```
I don't like this. Instead, we'll have the builder create a mapping and have CreateTable accept it.

praeclarum commented on src/SQLite.cs:
```diff
@@ -1946,77 +1941,440 @@ protected internal void Dispose()
+		}
+	}
+
+	public static class TableMappingFluentExtensions
```
These should not be public. I don't like introducing APIs unrelated to this library.

praeclarum commented on src/SQLite.cs:
```diff
@@ -1946,77 +1941,440 @@ protected internal void Dispose()
+
+	public class TableMappingBuilder<TEntity>
+	{
+		SQLiteConnection _sqlite;
```
I do not want a pointer back to the connection. I'm trying to keep Mappings separate from connections.

praeclarum commented on src/SQLite.cs:
```diff
@@ -1946,77 +1941,440 @@ protected internal void Dispose()
+		/// <returns>
+		/// The number of entries added to the database schema.
+		/// </returns>
+		public int CreateTable(CreateFlags createFlags = CreateFlags.None)
```
I want to remove this and instead introduce a `CreateTable(TableMapping map)` method that registers the mapping if needed but then hands off to the normal code.

praeclarum commented:
Great work!

praeclarum commented:
Tracking issue is #422

Eonasdan commented:
@praeclarum @RoyGoode what would it take to get this going? I started messing around with Roy's code based on your comments but this PR is from April has quite a bit of drift at this point.

sunthx commented:
Looking forward to this feature

RoyGoode commented:
Apologies to all in the delay in getting these updates in. I’ve spent pretty much the last 12 months working on non-.NET projects, but I’m back on Xamarin now for a new project. I will re-implement the Fluent API with the latest version and as per @praeclarum’s great suggestions.

RoyGoode commented:
As promised, I've updated to the latest source, and re-applied the new Fluent API changes, and also updated according to @praeclarum's reviews.

Here is an update to the example usage:

`var entity1 = TableMapping.Build<SampleEntity>()
		  .TableName("ExampleTable")
		  .PrimaryKey(x => x.Key)
		  .ToMapping();

var entity2 = TableMapping.Build<SampleEntity2>()
		  .TableName("Users")
		  .PrimaryKey(x => x.Id)
		  .AutoIncrement(x => x.Id)
		  .ColumnName(x => x.PasswordHash, "Password")
		  .Unique(x => x.Username)
		  .ToMapping();

var entity3 = TableMapping.Build<SampleEntity3>()
		  .TableName("Contacts")
		  .Index("ix_full", x => x.Name, x => x.Email, x => x.Telephone)
		  .PrimaryKey(x => x.Id, true)
		  .MaxLength(x => x.Email, 255)
		  .NotNull(x => x.Email)
		  .Ignore(x => x.SecretMessage)
		  .ToMapping();

connection.CreateTable(entity1);
connection.CreateTable(entity2);
connection.CreateTable(entity3);

// OR

connection.CreateTables(CreateFlags.None, entity1, entity2, entity3);`

RoyGoode commented:
I have created a fresh pull request #727 with a cleaner branch as my master was tracking the original master from circa April 2017.

Rivolvan commented:
this will be very helpful if this will be merged in the soonest possible time

rdvojmoc commented:
What is status on this pull request? This is a must have feature if you want to map third-party models.

Mrsevic commented:
Is there any status update available?

michaldobrodenka commented:
@Mrsevic I have merged and fixed current official and fluent api pull requests for testing, it seems ok. I'll try to use it in production.

lancecontreras commented:
Any update on this?
